docs/architecture/adr-047-handling-evidence-from-light-client.md

```diff
@@ -0,0 +1,165 @@
+
+### F4. Phantom validators
+
+A new type of evidence needs to be created:
+
+```go
+type PhantomValidatorEvidence struct {
+  PubKey crypto.PubKey
+  Vote types.Vote
+}
+```
+
+It contains a validator's public key and a vote for a block, where this
+validator is not part of the validator set.
+
+### F5. Lunatic validator
+
+```go
+type LunaticValidatorEvidence struct {
+  Header types.Header
+  Vote types.Vote
+}
+```
+
+To punish this attack, we need support for a new Evidence type -
+`LunaticValidatorEvidence`. This type includes a vote and a header. The header
+must contain fields that are invalid with respect to the previous block, and a
+vote for that header by a validator that was in a validator set within the
+unbonding period. While the attack is only possible if +1/3 of some validator
+set colludes, the evidence should be verifiable independently for each
+individual validator. This means the total evidence can be split into one piece
+of evidence per attacking validator and gossipped to nodes to be verified one
+piece at a time, reducing the DoS attack surface at the peer layer.
+
+Note it is not sufficient to simply compare this header with that committed for
+the corresponding height, as an honest node may vote for a header that is not
+ultimately committed. Certain fields may also be variable, for instance the
+`LastCommitHash` and the `Time` may depend on which votes the proposer includes.
+Thus, the header must be explicitly checked for invalid data.
+
+For the attack to succeed, VC must sign a header that changes the validator set
+to consist of something they control. Without doing this, they can not
+otherwise attack the light client, since the client verifies commits according
+to validator sets. Thus, it should be sufficient to check only that
+`ValidatorsHash` and `NextValidatorsHash` are correct with respect to the
+header that was committed at the corresponding height.
+
+That said, if the attack is conducted by +2/3 of the validator set, they don't
+need to make an invalid change to the validator set, since they already control
+it. Instead they would make invalid changes to the `AppHash`, or possibly other
+fields. In order to punish them, then, we would have to check all header
+fields.
+
+Note some header fields require the block itself to verify, which the light
+client, by definition, does not possess, so it may not be possible to check
+these fields. For now, then, `LunaticValidatorEvidence` must be checked against
+all header fields which are a function of the application at previous blocks.
+This includes `ValidatorsHash`, `NextValidatorsHash`, `ConsensusHash`,
+`AppHash`, and `LastResultsHash`. These should all match what's in the header
+for the block that was actually committed at the corresponding height, and
+should thus be easy to check.
+
+## Status
+
+Proposed.
+
+## Consequences
+
+### Positive
+
+* Tendermint will be able to detect & punish new types of misbehavior
+
```

Comment:
light clients connected to multiple full nodes can help full nodes notice a fork faster?